arelle/ModelRelationshipSet.py

```diff
@@ -204,50 +206,50 @@ def clear(self):
         if self.modelConceptRoots is not None:
             del self.modelConceptRoots[:]
         self.linkqname = self.arcqname = None
-        
+
     def __bool__(self):  # some modelRelationships exist
         return len(self.modelRelationships) > 0
-        
+
     @property
     def linkRoleUris(self):
         if self.modellinkRoleUris is None:
             self.modellinkRoleUris = set(modelRel.linkrole for modelRel in self.modelRelationships)
         return self.modellinkRoleUris
-    
+
     def loadModelRelationshipsFrom(self):
         if self.modelRelationshipsFrom is None:
             self.modelRelationshipsFrom = defaultdict(list)
             for modelRel in self.modelRelationships:
                 fromModelObject = modelRel.fromModelObject
                 if fromModelObject is not None: # none if concepts failed to load
                     self.modelRelationshipsFrom[fromModelObject].append(modelRel)
-    
+
     def loadModelRelationshipsTo(self):
         if self.modelRelationshipsTo is None:
             self.modelRelationshipsTo = defaultdict(list)
             for modelRel in self.modelRelationships:
                 toModelObject = modelRel.toModelObject
                 if toModelObject is not None:   # none if concepts failed to load
                     self.modelRelationshipsTo[toModelObject].append(modelRel)
-                
-    def fromModelObjects(self):
+
+    def fromModelObjects(self) -> defaultdict[Any, list] | None:
```

Comment:
IMO the fact that `self.modelRelationshipsFrom` is a `defaultdict` is really more of an internal class implementation detail that shouldn't leak out of the class through the type system (return type would just be dict). Also, this function will never return None.

Reply:
Oh, VS suggested it would return `None` or `defaultdict` and I believe the reason for that is it's being defined as `None` here:
https://github.com/Arelle/Arelle/blob/dbcf5fd3e89edc45c99481dcb4438b2158e27ee8/arelle/ModelRelationshipSet.py#L182

And `defaultdict` here:
https://github.com/Arelle/Arelle/blob/dbcf5fd3e89edc45c99481dcb4438b2158e27ee8/arelle/ModelRelationshipSet.py#L219

![image](https://user-images.githubusercontent.com/21142447/184194673-f28bdff8-b8e0-4b85-8460-0a038a9767dd.png)

Do we want to remove the definition on line 182? Mypy doesn't complain if I remove `None` as a return type in `fromModelObjects` though. I don't feel strongly about this through, I'm happy to remove `None` as the return type.

Reply:
`self.modelRelationshipsFrom` can be `None`, but never when it's the return value of this function. The call to `loadModelRelationshipsFrom` always sets it to a dict before this returns.

Reply:
I see what you mean now, thanks!